core/test/config/registry.cpp

```diff
@@ -0,0 +1,195 @@
+    auto found_solver_factory =
+        reg.search_data<gko::LinOpFactory>("solver_factory");
+    auto found_stop_factory =
+        reg.search_data<gko::stop::CriterionFactory>("stop_factory");
+
+    // get correct ptrs
+    ASSERT_EQ(found_matrix, matrix);
+    ASSERT_EQ(found_solver_factory, solver_factory);
+    ASSERT_EQ(found_stop_factory, stop_factory);
+    // get correct types
+    ASSERT_TRUE((std::is_same<decltype(found_matrix),
+                              std::shared_ptr<gko::LinOp>>::value));
+    ASSERT_TRUE((std::is_same<decltype(found_solver_factory),
+                              std::shared_ptr<gko::LinOpFactory>>::value));
+    ASSERT_TRUE(
+        (std::is_same<decltype(found_stop_factory),
+                      std::shared_ptr<gko::stop::CriterionFactory>>::value));
+}
+
+
+TEST_F(Registry, SearchDataWithType)
+{
+    reg.emplace("matrix", matrix);
+    reg.emplace("solver_factory", solver_factory);
+    reg.emplace("stop_factory", stop_factory);
+
+    auto found_matrix = reg.search_data<Matrix>("matrix");
+    auto found_solver_factory =
+        reg.search_data<Solver::Factory>("solver_factory");
+    auto found_stop_factory = reg.search_data<Stop::Factory>("stop_factory");
+
+    // get correct ptrs
+    ASSERT_EQ(found_matrix, matrix);
+    ASSERT_EQ(found_solver_factory, solver_factory);
+    ASSERT_EQ(found_stop_factory, stop_factory);
+    // get correct types
+    ASSERT_TRUE(
+        (std::is_same<decltype(found_matrix), std::shared_ptr<Matrix>>::value));
+    ASSERT_TRUE(
+        (std::is_same<decltype(found_solver_factory),
+                      std::shared_ptr<typename Solver::Factory>>::value));
+    ASSERT_TRUE((std::is_same<decltype(found_stop_factory),
+                              std::shared_ptr<typename Stop::Factory>>::value));
+}
+
+
+TEST_F(Registry, ThrowIfNotFound)
+{
+    ASSERT_THROW(reg.search_data<gko::LinOp>("N"), std::out_of_range);
+    ASSERT_THROW(reg.search_data<gko::LinOpFactory>("N"), std::out_of_range);
+    ASSERT_THROW(reg.search_data<gko::stop::CriterionFactory>("N"),
+                 std::out_of_range);
+}
+
+
+TEST_F(Registry, ThrowWithWrongType)
+{
+    reg.emplace("matrix", matrix);
+    reg.emplace("solver_factory", solver_factory);
+    reg.emplace("stop_factory", stop_factory);
+
+    ASSERT_THROW(reg.search_data<gko::matrix::Dense<double>>("matrix"),
+                 gko::NotSupported);
+    ASSERT_THROW(
+        reg.search_data<gko::solver::Cg<double>::Factory>("solver_factory"),
+        gko::NotSupported);
+    ASSERT_THROW(reg.search_data<gko::stop::Time::Factory>("stop_factory"),
+                 gko::NotSupported);
+}
+
+
+TEST_F(Registry, GetBuildMap)
+{
+    auto factory =
+        reg.get_build_map()
+            .at("func")(pnode{"unused"}, reg, type_descriptor{"void", "void"})
+            .on(exec);
+
+    ASSERT_NE(factory, nullptr);
+}
+
+
+TEST(TypeDescriptor, TemplateCreate)
+{
+    auto td1 = make_type_descriptor<double, int>();
```

Comment:
maybe put each type_descriptor into a separate section, eg
```c++
{
    SCOPED_TRACE("make with two parameters");
    auto td = make_type_descriptor<double, int>();

    ASSERT_EQ(td.get_value_typestr(), "double");
    ASSERT_EQ(td.get_index_typestr(), "int");
}
```
Otherwise, the list of `ASSERT_EQ(tdN, ...)` is a bit hard to follow